cloud/scope/powervs_cluster.go

```diff
@@ -157,3 +261,1558 @@ func (s *PowerVSClusterScope) PatchObject() error {
+	return nil
+}
+
+// SetLoadBalancerStatus set the loadBalancer id.
+func (s *PowerVSClusterScope) SetLoadBalancerStatus(name string, loadBalancer infrav1beta2.VPCLoadBalancerStatus) {
+	if s.IBMPowerVSCluster.Status.LoadBalancers == nil {
+		s.IBMPowerVSCluster.Status.LoadBalancers = make(map[string]infrav1beta2.VPCLoadBalancerStatus)
+	}
+	if val, ok := s.IBMPowerVSCluster.Status.LoadBalancers[name]; ok {
+		if val.ControllerCreated != nil && *val.ControllerCreated {
+			loadBalancer.ControllerCreated = val.ControllerCreated
+		}
+	}
+	s.IBMPowerVSCluster.Status.LoadBalancers[name] = loadBalancer
+}
+
+// GetLoadBalancerID returns the loadBalancer.
+func (s *PowerVSClusterScope) GetLoadBalancerID(loadBalancer infrav1beta2.VPCLoadBalancerSpec) *string {
+	if loadBalancer.Name == "" || s.IBMPowerVSCluster.Status.LoadBalancers == nil {
+		return nil
+	}
+	if val, ok := s.IBMPowerVSCluster.Status.LoadBalancers[loadBalancer.Name]; ok {
+		return val.ID
+	}
+	return nil
+}
+
+// GetLoadBalancerState will get the state for the load balancer.
+func (s *PowerVSClusterScope) GetLoadBalancerState(name string) *infrav1beta2.VPCLoadBalancerState {
+	if s.IBMPowerVSCluster.Status.LoadBalancers == nil {
+		return nil
+	}
+	if val, ok := s.IBMPowerVSCluster.Status.LoadBalancers[name]; ok {
+		return &val.State
+	}
+	return nil
+}
+
+// GetLoadBalancerHost will return the hostname of load balancer.
+func (s *PowerVSClusterScope) GetLoadBalancerHost(name string) *string {
+	if s.IBMPowerVSCluster.Status.LoadBalancers == nil {
+		return nil
+	}
+	if val, ok := s.IBMPowerVSCluster.Status.LoadBalancers[name]; ok {
+		return val.Hostname
+	}
+	return nil
+}
+
+// SetCOSBucketStatus set the COS bucket id.
+func (s *PowerVSClusterScope) SetCOSBucketStatus(cosInstanceID string, controllerCreated bool) {
+	if s.IBMPowerVSCluster.Status.COSInstance == nil {
+		s.IBMPowerVSCluster.Status.COSInstance = &infrav1beta2.ResourceReference{
+			ID:                &cosInstanceID,
+			ControllerCreated: &controllerCreated,
+		}
+	} else {
+		// set the controllerCreated only if its not set
+		if !*s.IBMPowerVSCluster.Status.COSInstance.ControllerCreated {
+			s.IBMPowerVSCluster.Status.COSInstance.ControllerCreated = &controllerCreated
+		}
+		s.IBMPowerVSCluster.Status.COSInstance.ID = &cosInstanceID
+	}
+}
+
+// GetResourceGroupID returns the resource group id.
+func (s *PowerVSClusterScope) GetResourceGroupID() (string, error) {
+	if s.IBMPowerVSCluster.Spec.ResourceGroup != nil && s.IBMPowerVSCluster.Spec.ResourceGroup.ID != nil {
+		return *s.IBMPowerVSCluster.Spec.ResourceGroup.ID, nil
+	}
+	if s.IBMPowerVSCluster.Status.ResourceGroup != nil && s.IBMPowerVSCluster.Status.ResourceGroup.ID != nil {
+		return *s.IBMPowerVSCluster.Status.ServiceInstance.ID, nil
+	}
+	resourceGroupID, err := s.fetchResourceGroupID()
+	if err != nil {
+		return "", err
+	}
+	s.SetStatus(ResourceGroup, infrav1beta2.ResourceReference{ID: &resourceGroupID, ControllerCreated: pointer.Bool(false)})
```

Comment:
why new setting this value in the Get method? 🤔

Reply:
We are expecting user to pass resource group name, So we only set the id. The GetResourceGroupID is called from various places but only for the first call the status will be set and in future it will be fetched from status itself.

Reply:
but its better to call that separately and set the status instead of in this function

Reply:
Updated.